### ISSUE TITLE
Add Support for Cardbook

### ISSUE THREAD

**Issue by GrantHorizons** [RealRaven2000/SmartTemplates#50]
I'm using CardBook extension as the default TB Address book is very limited - but there seems to be no way to get SmartTemplate4 to use CardBook. A "Which Address book to use?" question would be useful! And which address book IN Cardbook too.
thanks!

**Comment by RealRaven2000** [comment]
Please note that I am using an XPCOM function for retrieving AB data - it is not clear to me where this comes from. SmartTemplate uses the `Components.interfaces.nsIAbDirectory` - here is the relevant code:

```
  {
    let abManager = Components.classes["@mozilla.org/abmanager;1"].getService(Components.interfaces.nsIAbManager);
    allAddressBooks = abManager.directories; 
  }
  while (allAddressBooks.hasMoreElements()) {
    let addressBook = allAddressBooks.getNext()
                                     .QueryInterface(Components.interfaces.nsIAbDirectory);
    if (addressBook instanceof Components.interfaces.nsIAbDirectory) { // or nsIAbItem or nsIAbCollection
      // alert ("Directory Name:" + addressBook.dirName);
      try {
        let card = addressBook.cardForEmailAddress(mail);
        if (card)
          return card;

```

- as far as I know the interface `nsIAbDirectory` is able to retrieve data from CardBook, but you have to make sure not having a matching email address in the standard AB. I assume that cards in the standard AB have precedence when using the XPCOM method. XPCOM is going to be deprecated in favor of web APIs (which aren't written yet) - I would highly recommend filing a bug to include support for CardBook in the planned APIs. 

Thunderbird 78 is planned for release in June 2020, I am very doubtful all APIs will be replaced by that time, this means SmartTemplate4 will have to be rereleased as a "web experiment", which also requires a lot of rewriting. (realisticly between 4 and 8 man-weeks) - I am currently considering which route to take; probably a kickstarter to hire an external developer for the conversion work so that I can continue working on bugs, maintenance, new features and user support!

**Comment by RealRaven2000** [comment]
Going to pin this one for next implementation. Sorry for taking such a long time. It requires rewriting of pretty much all functions as "asynchronous".

**Comment by GrantHorizons** [comment]
It always seems so simple - but never is. Thanks for the update

**Comment by richanl** [comment]
I am also eagerly awaiting the integration of Cardbook fields into Smart template: #141

**Comment by RealRaven2000** [comment]
After such a long time, this still isn't implemented in 2023 - it's time this has to become a reality. I will need the help of Philippe (the author of Cardbook) in order to pull it off; I will ask him for some boilerplate code on our next Add-on Developer meeting (next Thursday).

**Comment by tkcantonedu** [comment]
I'd also love to see this integration. Both SmartTemplates and CardBook are powerful tools I use regularly.

**Comment by RealRaven2000** [comment]
Working on this right now, the very step is to make all calling functions async. which also means the main parser (in the experimental code) needs to become async. I have already done similar work in the (unfinished) web extension portion of the code (files starting with st-) so I am hopeful I can reuse some of that code.

**Comment by RealRaven2000** [comment]
List of functions that need to be async-ified - initial list as starting point:

```
getCardFromAB  - local function in mime.split, retrieves cards from the address book
SmartTemplate4.mimeDecoder.spli - this is doing all address field processing
Local functions [SmartTemplate4.regularize] 
simplify
classifyReservedWord
replaceReservedWords

SmartTemplate4.regularize()

classSmartTemplate.getProcessedText
extractSignature
classSmartTemplate.insertTemplate
insertFileEntryInComposer
loadIdentity
SmartTemplate4.init

smartTemplate_loadIdentity - this is a MonkeyPatch for the Thunderbird function LoadIdentity()


Problem with `classifyReservedWord`: this is passed as argument to a `string.replace` function - which by default expects a sync function. I believe I have already solved this very difficult problem in the mx code branch, just need to check how exactly it was done. I wrote a utility function `replaceAsync` in st-util.mjs.js - probably need to integrate this into my main SmartTemplate4.Util namespace.

this might be a major problem, (close to a showstopper) as Thunderbird surely calls this as a sync function. Need to check the Thunderbird source code, to see what it does "afterwards". Reminder - you can all an async function from a sync function, but you cannot "await" its result - the calling code will just continue relentlessly (even if LoadIdentity is still working or working later on).
Example for it being called by Thunderbird - during `NotifyComposeBodyReady()` it does some processing, then calls `loadHTMLMsgPrefs()` and `AdjustFocus()`
https://searchfox.org/comm-central/source/mail/components/compose/content/MsgComposeCommands.js#582

in `replaceReservedWords()`:
`modifyHeader`

**Comment by RealRaven2000** [comment]
next to asyncify:

`expand = function(str) { return str.replace(/%([\w-]+)%/gm, replaceReservedWords); };`

also

```
getQuoteHeader
getSmartTemplate
```

Result so far:
![image](https://user-images.githubusercontent.com/6314674/218260222-09c940b7-f19e-477d-99eb-430533281aec.png)

Still getting a "promise" output in resulting mail, so I am probably missing at least one await somewhere. Also the missing from / to headers may be a problem of call order, not sure yet. Are these headers available at time of replacement?

**Comment by RealRaven2000** [comment]
First test version with async processing - still some problems with address headers but it's a start:

[smartTemplate-fx-3.16pre31.zip](https://github.com/RealRaven2000/SmartTemplates/files/10713755/smartTemplate-fx-3.16pre31.zip)

once async works reliably I can start tests with Cardbook's new interfaces.

**Comment by RealRaven2000** [comment]
Asyncified code some more, see below. THis is before adding any of Phillipe's functions (CardBook).

[smartTemplate-fx-3.16pre36.zip](https://github.com/RealRaven2000/SmartTemplates/files/10714300/smartTemplate-fx-3.16pre36.zip)

Will probably commit this changeset to ESR91 - I will test this version on my production profile for a day or two before doing the actual Cardbook notifications. Might have to add more async;ified code before that.

**Comment by RealRaven2000** [comment]
Structure of a card returned by cardbook (below). Spelling & names obviously different from normal AB format. I am missing `displayName`. Need to check how to translate to the legacy format or vCard (unless this is in vCard format already)

```
adr: Array []
agent: ""
anniversary: ""
bday: ""
birthplace: ""
cacheuri: undefined
cardurl: ""
categories: Array []
cbid: "143b253f-d955-4326-b0dc-ab161e558074::28ff0dbc-1803-44a6-86be-b980ae41fa76"
class1: ""
created: false
deathdate: ""
deathplace: ""
deleted: false
dirPrefId: "143b253f-d955-4326-b0dc-ab161e558074"
email: Array [ (4) […] ]
emails: Array [ "thunderbirddaily@gmail.com" ]
etag: ""
firstname: ""
fn: "thunderbirddaily"
gender: ""
geo: ""
impp: Array []
isAList: false
key: ""
kind: ""
lastname: ""
logo: Object { types: [], value: "", localURI: "", … }
mailer: ""
member: Array []
nickname: ""
note: ""
org: ""
othername: ""
others: Array []
pg: Object {  }
pgname: ""
photo: Object { types: [], value: "", localURI: "", … }
prefixname: ""
prodid: "-//Thunderbird.net/NONSGML Thunderbird CardBook V28.8//EN"
rev: "20180511T195945Z"
role: ""
sortstring: ""
sound: Object { types: [], value: "", localURI: "", … }
suffixname: ""
tel: Array []
title: ""
tz: ""
uid: "28ff0dbc-1803-44a6-86be-b980ae41fa76"
updated: false
url: Array []
version: "4.0"
```

**Comment by RealRaven2000** [comment]
I just did a first check in after writing conversion code for Cardbook formats. Here is a first build that does something worth testing:

[smartTemplate-fx-3.16pre50.zip](https://github.com/RealRaven2000/SmartTemplates/files/10717356/smartTemplate-fx-3.16pre50.zip)

Added a special switch that needs to be activated in about:config. Just set
 
```
extensions.smartTemplate4.mime.resolveAB.CardBook = true
```

to activate Cardbook support. You also need the minimal version of 84.4 which implements the prototype version of card lookup which is used to extract the data from the other Add-on:

[cardbook-84.4-tb.zip](https://github.com/RealRaven2000/SmartTemplates/files/10717364/cardbook-84.4-tb.zip)

To install these preview versions, download both zip files and drag them into Thunderbird Add-ons Manager, without extracting their contents.

**Comment by RealRaven2000** [comment]
The changeset for going async:  [b3206cc](https://github.com/RealRaven2000/SmartTemplates/commit/b3206ccd3d4c60e04b20120579c5160798529293) 
shows which functions have been converted to asynchronous flow.

**Comment by RealRaven2000** [comment]
I added User Interface integration for Cardbook - find the new option to enable Cardbook on the advanced settings tab in SmartTemplates Settings:

![image](https://user-images.githubusercontent.com/6314674/218547955-6f8dd0f4-e668-4083-ab82-d37a724bfb51.png)

[smartTemplate-fx-3.16pre61.zip](https://github.com/RealRaven2000/SmartTemplates/files/10725335/smartTemplate-fx-3.16pre61.zip)

UI is already translated to the supported languages. I also added a better error message (and detailed loggin in error console) if any exceptions happen during replacing of SmartTemplates variables.

To install the above preview version download the zip files and drag it into Thunderbird Add-ons Manager, without extracting their contents. Note that you need at least Cardbook 84.4 installed (see [version above](url)) to test this new feature.

**Comment by RealRaven2000** [comment]
A new milestone was reached - with John Bieling's help I fixed the asynchronous replacement code which broke importing other templates via `%file()%` - Here is the latest version:

[smartTemplate-fx-3.16pre72.zip](https://github.com/RealRaven2000/SmartTemplates/files/10777684/smartTemplate-fx-3.16pre72.zip)

To install the above preview version download the zip files and drag it into Thunderbird Add-ons Manager, without extracting their contents. Note that you need at least Cardbook 84.4 installed (see [version above](https://github.com/RealRaven2000/SmartTemplates/issues/url)) to test this new feature.

**Comment by tkcantonedu** [comment]
Do I need the pro license to use it? I have the standard license.

**Comment by RealRaven2000** [comment]
> Do I need the pro license to use it? I have the standard license.

yes, this will be a premium feature, Pro License (or Domain license) is required. It's a very complex feature which required very fundamental rewriting + there is always the risk of depending on another Add-on and its potential changes.

If you want to wait until it lands officially in the next version 3.16 you can save some money on the upgrade as there will be a substantial bargain offer for upgraders when I launch it.

**Comment by RealRaven2000** [comment]
> Do I need the pro license to use it? I have the standard license.

PS: If you want to test now, you contact me directly for a test license.

**Comment by tkcantonedu** [comment]
Sounds good to me. I'll wait for the official release. Thanks!

**Comment by RealRaven2000** [comment]
Here is a test version with some interesting caveats. This parses the Cardbook structure directly. But I also improved the vCard parser (which is the new format used by Thunderbird standard AB) The way I parsed the vCard structure directly wasn't optimal as there are different version numbers which are implemented differently and could lead to failures. In this version I am using the ical library to parse out a jCal (json) structure, that hopefully is version independent. Since the CardBook part is parsing the native structure I will obviously need a heads up by the author in case he wants to modify the internal structure of his card storage.

[smartTemplate-fx-3.16pre109.zip](https://github.com/RealRaven2000/SmartTemplates/files/10882202/smartTemplate-fx-3.16pre109.zip)

Test template with all supported fields
[AB.zip](https://github.com/RealRaven2000/SmartTemplates/files/10882224/AB.zip)

**Comment by RealRaven2000** [comment]
Another version - adds improvements in vCard reading and CardBook for organisation / departements. You can now have multiple departments after the Organisation by delimiting them with `;` - these will automatically delimited with `<br>`. I am also contemplating whether to add an option "fallback to standard AB" if no entry is found in Cardbook.

[smartTemplate-fx-3.16pre115.zip](https://github.com/RealRaven2000/SmartTemplates/files/10885206/smartTemplate-fx-3.16pre115.zip)

To install, download the zip file and drag it into Thunderbird Add-ons Manager.

**Comment by RealRaven2000** [comment]
New test template:

[AB.zip](https://github.com/RealRaven2000/SmartTemplates/files/10885235/AB.zip)

unpack the html file and add to the "reply" menu (configure menu item) or just open it directly:

![image](https://user-images.githubusercontent.com/6314674/222800864-c139daef-e16b-4b14-a23a-c0c3741191de.png)

**Comment by RealRaven2000** [comment]
Latest version - added a "fallback to standard AB option" when contact is not found in Cardbook:

![image](https://user-images.githubusercontent.com/6314674/222912745-103eb1c6-1d57-413c-92e5-89a175f0c16a.png)


[smartTemplate-fx-3.16pre126.zip](https://github.com/RealRaven2000/SmartTemplates/files/10888909/smartTemplate-fx-3.16pre126.zip)

To try out, download the zip file and drag it into Thunderbird Add-ons Manager (don't extract contents).

**Comment by RealRaven2000** [comment]
implemented  in version 3.16 published on 6th of March 2023

**Comment by tkcantonedu** [comment]
Hi, I upgraded to the pro version and have used it for several weeks. It works great. Thanks!
